// playcanvas/build/playcanvas.dbg/src/platform/graphics/webgpu/webgpu-draw-commands.js
import { BUFFERUSAGE_INDIRECT, BUFFERUSAGE_COPY_DST } from '../constants.js';
import { StorageBuffer } from '../storage-buffer.js';

/**
 * @import { GraphicsDevice } from '../graphics-device.js'
 */ /**
 * WebGPU implementation of DrawCommands.
 *
 * @ignore
 */ class WebgpuDrawCommands {
    /**
     * Allocate AoS buffer and backing storage buffer.
     * @param {number} maxCount - Number of sub-draws.
     */ allocate(maxCount) {
        // Skip reallocation if size matches exactly
        if (this.gpuIndirect && this.gpuIndirect.length === 5 * maxCount) {
            return;
        }
        this.storage?.destroy();
        this.gpuIndirect = new Uint32Array(5 * maxCount);
        this.gpuIndirectSigned = new Int32Array(this.gpuIndirect.buffer);
        this.storage = new StorageBuffer(this.device, this.gpuIndirect.byteLength, BUFFERUSAGE_INDIRECT | BUFFERUSAGE_COPY_DST);
    }
    /**
     * Write a single draw entry.
     * @param {number} i - Draw index.
     * @param {number} indexOrVertexCount - Count of indices/vertices.
     * @param {number} instanceCount - Instance count.
     * @param {number} firstIndexOrVertex - First index/vertex.
     * @param {number} baseVertex - Base vertex (signed).
     * @param {number} firstInstance - First instance.
     */ add(i, indexOrVertexCount, instanceCount, firstIndexOrVertex, baseVertex = 0, firstInstance = 0) {
        const o = i * 5;
        this.gpuIndirect[o + 0] = indexOrVertexCount;
        this.gpuIndirect[o + 1] = instanceCount;
        this.gpuIndirect[o + 2] = firstIndexOrVertex;
        this.gpuIndirectSigned[o + 3] = baseVertex;
        this.gpuIndirect[o + 4] = firstInstance;
    }
    /**
     * Upload AoS data to storage buffer.
     * @param {number} count - Number of active draws.
     * @returns {number} Total primitive count.
     */ update(count) {
        if (this.storage && count > 0) {
            const used = count * 5; // 5 uints per draw
            this.storage.write(0, this.gpuIndirect, 0, used);
        }
        // calculate total primitives for stats
        let totalPrimitives = 0;
        if (this.gpuIndirect && count > 0) {
            for(let d = 0; d < count; d++){
                const offset = d * 5;
                const indexOrVertexCount = this.gpuIndirect[offset + 0];
                const instanceCount = this.gpuIndirect[offset + 1];
                totalPrimitives += indexOrVertexCount * instanceCount;
            }
        }
        return totalPrimitives;
    }
    destroy() {
        this.storage?.destroy();
        this.storage = null;
    }
    /**
     * @param {GraphicsDevice} device - Graphics device.
     */ constructor(device){
        /** @type {Uint32Array|null} */ this.gpuIndirect = null;
        /** @type {Int32Array|null} */ this.gpuIndirectSigned = null;
        /**
     * @type {StorageBuffer|null}
     */ this.storage = null;
        this.device = device;
    }
}

export { WebgpuDrawCommands };
